site/bps/BP-35-128-bits-support.md

```diff
@@ -0,0 +1,399 @@
+`ledgerQualifiedName` to `ledgerScopeId` and `ledgerId`.
+
+### Public Interfaces
+
+#### API Change
+
+The API will be introducing `ledgerScopeId` across the interfaces. This field will be optional and default to `0`. 
+
+##### Handle
+
+Introduce a new method `getScopeId` for representing the scope id (the most significant  `128 bits` ledger id).
+
+```java
+public interface Handle extends AutoCloseable {
+
+  ...
+
+  /**
+   * Return the ledger scope id. The most significant 64 bits of 128 bits.
+   */
+  long getScopeId();
+
+  /**
+   * Return the ledger id. The least significant 64 bits of 128 bits.
+   */ 
+  long getId();
+
+  ...
+
+}
+```
+
+##### Create LedgerAdv
+
+Introduce a new method `withLedgerScopeId` in `CreateAdvBuilder` for providing `scopeId`
+(the most significant 64 bits for 128 bits ledger id) on creating a ledger.
+
+```java
+public interface CreateAdvBuilder extends OpBuilder<WriteHandle> {
+  ...
+
+  /**
+   * Set the scope id for the newly created ledger.
+   * If no explicit scopeId is passed, the new ledger
+   * will be created under scope `0`.
+   */
+  CreateAdvBuilder withLedgerScopeId(long scopeId);	
+
+  ...
+}
+```
+
+##### Open Ledger
+
+Introduce a new method `withLedgerScopeId` in `OpenBuilder` for providing `scopeId`
+(the most significant 64 bits for 128 bits ledger id) on opening a ledger.
+
+```java
+public interface OpenBuilder extends OpBuilder<ReadHandle> {
+  ...
+  /**
+   * Set the scope id of the ledger to open.
+   */
+  OpenBuilder withLedgerScopeId(long scopeId);
+  ...
+}
+```
+
+##### Delete Ledger
+
+Introduce a new method `withLedgerScopeId` in `DeleteBuilder` for providing `scopeId`
+(the most significant 64 bits for 128 bits ledger id) on deleting a ledger.
+
+```java
+public interface DeleteBuilder extends OpBuilder<Void> {
+  ...
+  /**
+   * Set the scope id of the ledger to delete.
+   */
+  DeleteBuilder withLedgerScopeId(long scopeId);
+  ...
+}
+```
+
+#### CLI
+
+All BookKeeper CLI tools will be updated with additional option `—ledger-scope-id`.
+Optionally we can add option `—ledger-qualified-name` (the hex representation of 128 bits).
+Internally all the CLI tools will convert ledger qualified name to `ledgerId` and `ledgerScopeId`.
+
+#### REST
+
+1. All ledger related endpoints will be adding a new parameter `ledger_scope_id`. 
+2. `ListLedgerService`  only supports listing ledgers under a given ledger scope id.
+   If `ledger_scope_id` is missing, it will be listing ledgers under scope `0`.
+
+#### Wire Protocol
+
+> There will be no plan for supporting 128 bits in v2 protocol, due to the limitation in v2 protocol.
+> So any operations in v2 protocol with scope id not equal to 0  will be failed immediately with
+> `ILLEGAL_OP` exceptions.
+
+All the request and response messages will be adding an optional field `optional int64 ledgerScopeId`.
+
+#### Entry Format
+
+Currently all the entries written to bookies are encoded in a certain format, including `metadata`,
```

Comment:
it is not just wire protocol or journal format. it is the format of an entry:

- client is aware of this format (for serializing/deserializing an entry)
- it is transparent to wire protocol. so wire protocol doesn't care.
- journal needs to be aware of this format (for learning the header part, ledger id, entry id, lac and such).
- ledger storage needs to be aware of this format (same as journal).

that's why I kept this in its own section called "Entry Format".

Reply:
uff, this is going to touch a lot of places. When implementing, please make sure the v1 vs v2 stuff is well encapsulated, so that we don't end up with a load of ```if (v1) { doSomething } else if (v2) { doAnotherThing }```